src/hotspot/share/opto/loopPredicate.cpp

```diff
@@ -42,20 +42,161 @@
+ *                                                   (We could still only use one Template Assertion Predicate to create
+ *                                                   both Initialized Assertion Predicates from - might be worth doing
+ *                                                   at some point).
+ *                                                   When later splitting a loop (pre/main/post, peeling, unrolling),
+ *                                                   we create two Initialized Assertion Predicates from the Template
+ *                                                   Assertion Predicates by replacing the OpaqueLoop* nodes by actual
+ *                                                   values. Initially (before unrolling), both Assertion Predicates are
+ *                                                   equal. The Initialized Assertion Predicates are always true because
+ *                                                   their range is covered by a corresponding Hoisted Predicate.
+ *                        - Range Check Elimination: A range check is removed from the main-loop by changing the pre
+ *                                                   and main-loop iterations. We add two additional Template Assertion
+ *                                                   Predicates (see explanation in section above) and one Initialized
+ *                                                   Assertion Predicate for the just removed range check. When later
+ *                                                   unrolling the main-loop, we create two Initialized Assertion
+ *                                                   Predicates from the Template Assertion Predicates by replacing the
+ *                                                   OpaqueLoop* nodes by actual values for the unrolled loop.
+ *                                                   The Initialized Assertion Predicates are always true because their
+ *                                                   range is covered by the main-loop entry guard.
+ *
+ *
+ * In order to group predicates and refer to them throughout the code, we introduce the following additional terms:
+ * - Runtime Predicate Block: A block containing all Runtime Predicates belonging to a single Parse Predicate. This block
+ *                            could be empty if there were no Runtime Predicates created at this Parse Predicate.
+ *                            For the time being: We also count Assertion Predicates to this block but that will be
+ *                            changed with the redesign of Assertion Predicates where we remove them from this block
+ *                            (JDK-8288981).
+ * - Regular Predicate Block: A Regular Predicate Block consists of a Parse Predicate a Runtime Predicate Block (all its
+ *                            Runtime Predicates, if any). There are three such blocks:
```

Comment:
A Regular Predicate Block consists of a Runtime Predicate Block (all its Runtime Predicates, if any) with a Parse Predicate after it.